lib/notifiers/thomas2.dart

```diff
@@ -1,217 +1,90 @@
-        if (entry == _notifyLimit) {
-          _notifyLimit = _notifyLimit?.previous;
-
-          if (entry == _notifyCursor) {
-            _notifyCursor = _notifyLimit;
-          }
+    for (int i = 0; i < _listeners.length; i++) {
+      final _Listener _listener = _listeners[i];
+      if (_listener.func == listener && _listener.active) {
+        if (_notificationCallStackDepth > 0) {
+          _listener.active = false;
+          _removedListeners++;
         } else {
-          if (entry == _notifyCursor) {
-            _notifyCursor = _notifyCursor?.previous;
-          }
+          _listeners.removeAt(i);
         }
-        entry.unlink();
-        return;
+        break;
       }
     }
   }
 
-  /// Discards any resources used by the object. After this is called, the
-  /// object is not in a usable state and should be discarded (calls to
-  /// [addListener] and [removeListener] will throw after the object is
-  /// disposed).
-  ///
-  /// This method should only be called by the object's owner.
-  @mustCallSuper
   void dispose() {
-    assert(_debugAssertNotDisposed());
     _listeners = null;
   }
 
-  /// Call all the registered listeners.
-  ///
-  /// Call this method whenever the object changes, to notify any clients the
-  /// object may have changed. Listeners that are added during this iteration
-  /// will not be visited. Listeners that are removed during this iteration will
-  /// not be visited after they are removed.
-  ///
-  /// Exceptions thrown by listeners will be caught and reported using
-  /// [FlutterError.reportError].
-  ///
-  /// This method must not be called after [dispose] has been called.
-  ///
-  /// Surprising behavior can result when reentrantly removing a listener (e.g.
-  /// in response to a notification) that has been registered multiple times.
-  /// See the discussion at [removeListener].
-  @protected
-  @visibleForTesting
   void notifyListeners() {
-    assert(_debugAssertNotDisposed());
-    if (_listeners.isEmpty) return;
-
-    _notifyLimit = _listeners?.last;
-    _notifyCursor = _listeners?.first;
-    do {
-      try {
-        _notifyCursor?.listener();
-      } catch (exception, stack) {
-        FlutterError.reportError(FlutterErrorDetails(
-          exception: exception,
-          stack: stack,
-          library: 'foundation library',
-          context: ErrorDescription(
-              'while dispatching notifications for $runtimeType'),
-          informationCollector: () sync* {
-            yield DiagnosticsProperty<Thomas2ChangeNotifier>(
-              'The $runtimeType sending notification was',
-              this,
-              style: DiagnosticsTreeStyle.errorProperty,
-            );
-          },
-        ));
+    if (_listeners.isEmpty) {
+      return;
+    }
+    _notificationCallStackDepth++;
+
+    if (_listeners != null) {
+      _listenerLengh ??= _listeners.length;
+      for (int i = 0; i < _listenerLengh; i++) {
```

Comment:
By doing this, we prevent listeners added during the first iteration to be called in the second iteration. We would want to remove this instance field and use a variable instead.

Reply:
actually I did this for exactly this reason. I'm not sure if it is a good idea to have a recursive call calling the newly added, but I honestly think we should ass an assert that forbids recursive calls

Reply:
But by doing this, we could break existing code, I'm not sure we should change the current behavior.

Reply:
I thought the old Notifier didn't support recursion at all?

Reply:
I'm not sure, but since the list was cloned, I think it was possible to do it.

Reply:
Can we check this? I think @knaeckeKami or you said they did not support recursion. If so we should check that we have the same behaviour